### ISSUE TITLE
feat(auth): Integrate OAuth 2.0 as a provider strategy

### ISSUE THREAD

**Issue by raisedadead** [freeCodeCamp/freeCodeCamp#15914]
## Proposal:

Going ahead, we would want our users to sign up just once and be able to use their profile on all of our platforms. This is not the case every time.

This can be easily solved by implementing  OAuth 2.0 as a provider strategy.

**Note: This should definitely be considered only after we upgrade to LoopBack 3**

### Case 1: Forum

We could get rid of the flow in our forum onboarding, and hence reduce the spam/fake accounts by officially linking our current user base to the forum.

This reduces the friction, and users having to onboard once again.

That said here are some considerations and would have to be done in phases:
- [ ] We should take care of current users to be able to link to their existing freeCodeCamp accounts.
- [ ] Identify and fixup accounts that can't be linked.
- [ ] We should be then able to remove GitHub and email sign-ups from discourse.
- [ ] We allow only verified profiles to be able to access the forum via login.

### Case 2: Open API Management Console

As we plan to provide a robust API around the vast datasets that we have, authentication would be a key factor in providing access control via keys (Kind of like AWS IAM does) and hence rate limiting the usage as we consider the cost of being able to provide all of the data in open while being able to afford that economically and securely.

We would definitely be looking for a developer dashboard, and users should be able to request keys using this strategy.

### Case 3: Third party application integration

This is an extension of Point 2, and potentially can open up integration possibilities in ways that can be super interesting, for example, if someone creates a platform or already has one, for instance, a job board, they can potentially integrate our auth strategy to let users quickly apply using the freeCodeCamp user profiles.

They can even validate our certifications.

The possibilities are endless.

### Case X: TBD

To be added to more research and ideas go into this thread.

**Comment by raisedadead** [comment]
@freeCodeCamp/moderators I would love to hear your feedback on this proposal.

**Comment by dhcodes** [comment]
I love it.

**Comment by QuincyLarson** [comment]
@raisedadead First of all, I love the idea of unifying everything behind the same auth. This said, I don't think spam is that big of a problem on the forum. 

> We should take care of current users to be able to link to their existing freeCodeCamp accounts.
>  Identify and fixup accounts that can't be linked.

This would involve merging about 200,000 forum accounts, so all of this would need to be done programmatically. There could be no room for ambiguity - otherwise thousands of people would think "freeCodeCamp deleted my forum account" and there wouldn't be a good way to convince them otherwise.

So as much as I like the idea of doing this, it should be a super low priority.

Your second and third cases are much more relevant. 

Our current big feature roadmap looks something like this:
1. Ship beta
2. Upgrade to the latest version of Loopback
3. Implement open API to support ecosystem of external apps
4. Serve the platform completely as a static app (using Netlify)
5. Streamline performance with Redis for challenge solution storage (like JSBin uses)

Having an authentication service would help us with steps 3 and 4. Then we could talk about using it to authenticate people into the forum. I know that people have successfully done this with Discourse. I'm just not sure how much work it will take. I'd be interested in your take one this.

@BerkeleyTrue what are your thoughts on this?

**Comment by raisedadead** [comment]
@QuincyLarson I absolutely agree, to your points, and yes you have got the assumptions and concerns spot on.

Implementing the OAuth strategy for Discourse is fairly well documented in their meta documentation. I can assist as soon as I am done with my ongoing tasks with the LoopBack upgrade and stable passwordless auth.

Will, of course, keep you posted.

**Comment by QuincyLarson** [comment]
@raisedadead Awesome! I'm glad to hear you've already looked into this. I'm definitely open to doing it if you feel up for it and we can come up with a good way to merge all the records.

**Comment by BerkeleyTrue** [comment]
In the interest in keeping open issue count low and seeing how we have many projects that need to come to fruition before this issue can be worked on, I suggest we close this issue and reopen once those other projects are complete

**Comment by raisedadead** [comment]
Sure thing, I am closing this for now and adding it to ToDo in Open API project. Thanks.

**Comment by SenamelaMabotja** [comment]
How can we manage the courses better, because whenever I get back on my
training, I am required to start from scratch.

**Comment by raisedadead** [comment]
@SenamelaMabotja please request support on the forum https://forum.freeCodeCamp.org, this tracker is not for support requests.
